File: common/src/main/java/net/kimleo/rec/concept/Queryable.java
package net.kimleo.rec.concept;

import java.util.function.Function;

public interface Queryable<T> {

    default Queryable<T> select(QuerySelector<T> selector) {
        return selector.findAll(this);
    }

    default T select1(QuerySelector<T> selector) {
        return selector.findFirst(this);
    }

    Queryable<T> where(QuerySelector<T> selector, Function<T, Boolean> fn);
}

File: common/src/main/kotlin/net/kimleo/rec/concept/Queryable.kt
package net.kimleo.rec.concept

interface Queryable<T> {
    fun select(selector: QuerySelector<T>): Queryable<T> = selector.findAll(this)
    fun select1(selector: QuerySelector<T>): T? = selector.findOne(this)
    fun where(selector: QuerySelector<T>, fn: T.() -> Boolean): Queryable<T>
}